FINERACT-724 Upgrade Spring Boot, Spring and Spring Security to their latest stable version

ivbond opened the pull request:
This is upgrade of Spring Boot, Spring, Spring Security, OpenJpa and other related libraries to their latest versions. 
All Integrations tests are passing with the new commits. But I would like to describe issues I have encountered during development:
1) Spring security does not support custom SaltSource. it is now random generated. So I used MessageDigestPasswordEncoder and V353__migrate_passwords_to_ss_5.sql for password migration.
2) There is huge trobule with OpenJpa and Spring5 compatibility. OpenJpaVendorAdapter was removed from Spring5. (See:  https://jira.spring.io/browse/SPR-16035?redirect=false&focusedCommentId=160186&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel) So I have upgraded OpenJpa and moved some code from spring 4 to fineract project. This solution has to be reviewed twice. I do not like it but this is the only way I have made it works. I have tried without jpaVendorAdapter but it leads to:
`"SQLException: Lock wait timeout exceeded; try restarting transaction"`

3) Alot of issues I had with dependencies.gradle and dev-dependencies.gradle they have alot of duplicates and dev is out of date. I think after Tomcat removal we could will leave only dependencies.gradle.

conradsp committed:
Set host to fineractmysql when running in docker

marta-jankovics committed:
Openbanking - API Gateway 1.0

dial-registry committed:
Merge remote-tracking branch 'upstream/develop' into develop

dial-registry committed:
Merge remote-tracking branch 'marta/develop' into develop

dial-registry committed:
Fixes for license headers

dial-registry committed:
Add healthcheck so DB is up before Fineract starts

ivanbonzl committed:
implementation of FINERACT-724 (Upgrade Spring Boot, Spring and Spring Security to their latest stable version)

ivanbonzl committed:
implementation of FINERACT-724 (Upgrade Spring Boot, Spring and Spring Security to their latest stable version)

ivanbonzl committed:
implementation of FINERACT-724 properties cleanup

ivanbonzl committed:
implementation of FINERACT-724 properties cleanup

vorburger commented:
WOW! Real nice. -- I'll try to find time to review this properly ASAP; say coming 1 week. (The security related changes in particular probably deserve a closer look.)

vorburger committed:
Add Z,ro flags to Docker Compose Volume

This is required on secure Linux distributions such as
Fedora/CentOS/RHEL where SELinux is enabled, where without this
the mysql container fails to start due to:

  ls: cannot open directory '/docker-entrypoint-initdb.d/': Permission denied

These Z,ro volume mount flags are (should be..) harmless and ignored
by Docker on other platforms, such as insecure ;) Linux distributions
without SELinux, or Windows and Mac.

PS: I personally don't actually use real Docker, but instead I use Podman
(see https://github.com/containers/libpod; install via `dnf install podman-docker`)
and https://github.com/containers/podman-compose/) via `pip3 install podman-compose`;
it's more secure (because it doesn't need a daemon and root to run containers).

vorburger committed:
Add docker-compose non-regression test to Travis CI

vorburger committed:
docker-compose.yml version 3.3 instead of 3.7

conradsp committed:
Merge pull request #643 from vorburger/docker-compose-selinux

Add Z,ro flags to Docker Compose Volume

conradsp committed:
Merge pull request #640 from conradsp/develop

FINERACT-794 Open Banking APIs for Interoperability layer

vorburger committed:
Minor doc improvements re. Docker in README

vorburger commented:
@conradsp @awasum @Anh3h @vishwasbabu have you seen this one?  Really cool, no?

Perhaps any of you'd like to chime in & help review this as well?

vorburger committed:
Stop mysql system service before docker-compose on Travis CI

vorburger committed:
install httpie in .travis.yml for docker-compose test

vorburger committed:
sleep 30s instead of 5s before attempting to httpie

vorburger commented on fineract-provider/build.gradle:
```diff
@@ -21,15 +21,15 @@ Run as:
+project.ext.springOauthVersion = '2.3.6.RELEASE'
 project.ext.jerseyVersion = '1.17'
-project.ext.springDataJpaVersion = '1.7.0.RELEASE' // also change spring-boot-gradle-plugin version below
+project.ext.springDataJpaVersion = '2.2.0.RELEASE' // also change spring-boot-gradle-plugin version below
```
FTR: I've verified that all these are the latest current versions; except for springVersion where there's already a 5.2.0 whereas this mvoes us up to 5.1.10; may be could bump to latest in a follow-up PR? Re. Spring OAuth, I'm not sure (doubt) if there are any integration tests covering that, so in theory this could break it.

vorburger commented on fineract-provider/dependencies.gradle:
```diff
@@ -66,7 +66,7 @@ dependencies {
                 //[group: 'mysql', name: 'mysql-connector-java', version: '5.1.27'],
                 [group: 'org.drizzle.jdbc', name: 'drizzle-jdbc', version: '1.4'],
-               // [group: 'org.apache.tomcat', name: 'tomcat-jdbc', version: tomcatVersion],
+                [group: 'org.apache.tomcat', name: 'tomcat-jdbc', version: tomcatVersion],
```
@ivan333m could you let us know why this now required?

ivbond replied:
New version of spring-boot-starter-data-jpa uses HikariCP as default connection pool instead of tomcat-jdbc. So I need to add this dependency. Also I have excluded HikariCP in the latest commit.

vorburger replied:
OK, thanks for the clarification. I've opened https://issues.apache.org/jira/browse/FINERACT-796 suggesting that someone (you?) could replace Tomcat JDBC with HikariCP later.

vorburger commented on fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/common/Utils.java:
```diff
@@ -61,7 +61,7 @@
     public static final String TENANT_TIME_ZONE = "Asia/Kolkata";
 
-    private static final String HEALTH_URL = "/fineract-provider/health";
+        private static final String HEALTH_URL = "/fineract-provider/actuator/health";
```
ivan333m@ this extra indent is probably a mistake, do you want to remove it?

vorburger replied:
@ivan333m if you are looking to contribute :smile: more, this reminded me of https://issues.apache.org/jira/browse/FINERACT-712 ...

vorburger commented on fineract-provider/dev-dependencies.gradle:
```diff
@@ -17,7 +17,7 @@
  */
 dependencies {
-        def tomcatVersion = '7.0.54'
+        def tomcatVersion = '7.0.94'
```
Great! This aligns with `bitnami/tomcat:7.0.94` in our `Dockerfile`.

vorburger commented on fineract-provider/dev-dependencies.gradle:
```diff
@@ -27,22 +27,19 @@ dependencies {
 
-				[group: 'org.apache.openjpa', name:'openjpa-all', version:'2.4.1'],
-                [group: 'org.apache.openjpa', name:'openjpa-maven-plugin', version:'2.4.1'],
+                [group: 'org.apache.openjpa', name:'openjpa-all', version:openJPAVersion],
```
I remember (going through some pain to) remove `openjpa-all` in https://github.com/apache/fineract/pull/607/files, and am surprised to see it still appear here. I know this just changes the version, but would you like to check out if this could just be removed all together?

ivbond replied:
`openjpa-all` has been removed

vorburger replied:
Thanks! PS: I know the `dev-dependencies.gradle` is a mess.. do you want to help to clean it up?

vorburger commented on fineract-provider/dev-dependencies.gradle:
```diff
@@ -67,7 +64,7 @@ dependencies {
 
                 //[group: 'mysql', name: 'mysql-connector-java', version: '5.1.27'],
-                [group: 'org.drizzle.jdbc', name: 'drizzle-jdbc', version: '1.3'],
+                [group: 'org.drizzle.jdbc', name: 'drizzle-jdbc', version: '1.4'],
```
as in https://github.com/apache/fineract/pull/616, perfect!

vorburger commented on fineract-provider/dev-dependencies.gradle:
```diff
@@ -92,17 +89,19 @@ dependencies {
                 [group: 'io.swagger', name: 'swagger-jersey-jaxrs', version: '1.5.15'],
-                [group: 'org.springframework', name:'spring-jms'],
-                [group: 'org.apache.activemq', name: 'activemq-broker'],
+                [group: 'org.springframework', name:'spring-jms', version: '4.0.7.RELEASE'],
```
In an ideal world, in a future PR, it would be really really cool if we could just avoid having to repeat using fixed version, and could use the implicit versions via Spring BOM dependency management, see https://docs.spring.io/spring-boot/docs/current/gradle-plugin/reference/html/#managing-dependencies ...

ivbond replied:
It is really nice idea. But I would propose to move it into dedicated task, just to keep the current task shorter.

vorburger replied:
Of course, I didn't mean to do that in this PR, but later; FYI I've just created https://issues.apache.org/jira/browse/FINERACT-797 re. this.. just in case you'd like to pick that up some day! :smiling_imp:

vorburger commented on fineract-provider/properties/basicauth/application.properties:
```diff
@@ -17,4 +17,5 @@
-spring.profiles.default=basicauth
\ No newline at end of file
+spring.profiles.default=basicauth
+management.health.jms.enabled=false
```
@ivan333m could you let us know why this is now required? It seems to disable "Health Management" (JMX ?) for JMS... I'm guessing you had a good reason to have to do this, so I'm sure it's OK to leave it like this - just let us know why this is required?

ivbond replied:
Unfortunately without this property I have a lot warning that JMS is not available. 

`10:53:03.826 [http-bio-8443-exec-2] WARN  o.s.b.actuate.jms.JmsHealthIndicator - JMS health check failed
javax.jms.JMSException: Could not connect to broker URL: tcp://localhost:61616. Reason: java.net.ConnectException: Connection refused (Connection refused)
        at org.apache.activemq.util.JMSExceptionSupport.create(JMSExceptionSupport.java:36)`

vorburger replied:
OK, thanks for the explanation.

vorburger commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/boot/AbstractApplicationConfiguration.java:
```diff
@@ -48,7 +52,11 @@
+		GsonAutoConfiguration.class,
+		JdbcTemplateAutoConfiguration.class})
+@EnableWebSecurity
+@EnableTransactionManagement
```
@ivan333m how come `@EnableWebSecurity` and `@EnableTransactionManagement` are required here now? Were they implicit in older Spring versions, and need to be made explicit in new versions?

ivbond replied:
@EnableTransactionManagement is a replacement for <tx:annotation-driven /> . Without @EnableWebSecurity we need to configure additional spring security beans for example `org.springframework.security.config.annotation.ObjectPostProcessor`

vorburger replied:
OK, so looks like the default changed in the newer Spring version. This couldn't hurt, so fine.

vorburger commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/boot/WebXmlConfiguration.java:
```diff
@@ -46,10 +47,12 @@
     public Filter springSecurityFilterChain() {
         return new DelegatingFilterProxy();
     }
+*/
```
for future clarify better remove than only comment out... but I'd love more eyes on this particular change - this is security related - why is this Filter not required anymore in newer Spring versions?

ivbond replied:
Yep, I have such problem. I will do final self-review before commit more carefully.

vorburger replied:
I saw that now it's cleanly removed instead of being commented out, but I still don't really understand why this `DelegatingFilterProxy` is not required anymore... particularly as this seems to be security related, I'm still curious to have a record of why?

ivbond replied:
It is `DelegatingFilterProxy` duplicate. Spring security will create it by default.

vorburger commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/domain/DefaultPasswordEncoderFactories.java:
```diff
@@ -0,0 +1,54 @@
+import java.util.Map;
+
+
+public class DefaultPasswordEncoderFactories {
```
Where is this code from? I think a JavaDoc would be very useful here, may be something à la "This class is a verbatim copy/paste from Spring vX, in order to keep backwards compatibility in Fineract, because it was removed from Spring vY." (if that is what happened here? if it's for another reason, likewise explain in JavaDoc useful for future archeologists)

ivbond replied:
Sorry. But it was not copy/pasted from Spring. I have found parts of this code in the Spring security migration documentation and forums.

vorburger commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/service/PaginationHelper.java:
```diff
@@ -31,8 +32,8 @@
         // determine how many rows are available
-        @SuppressWarnings("deprecation")
-        final int totalFilteredRecords = jt.queryForInt(sqlCountRows);
+//        final int totalFilteredRecords = jt.queryForInt(sqlCountRows);
```
remove instead of just commenting out un-used code (it's clearer for future readers)

ivbond replied:
Done

vorburger commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadWriteNonCoreDataService.java:
```diff
@@ -32,16 +32,16 @@
     DatatableData retrieveDatatable(String datatable);
 
-    @PreAuthorize(value = "hasAnyRole('ALL_FUNCTIONS', 'REGISTER_DATATABLE')")
+    @PreAuthorize(value = "hasAnyAuthority('ALL_FUNCTIONS', 'REGISTER_DATATABLE')")
```
@ivan333m why this change? I don't know much about Spring Security myself, but looking at https://docs.spring.io/spring-security/site/docs/current/reference/htmlsingle/ it seems that Role and Authority are not the same, and both are still valid in current versions? -- NB: This change also appears in other files below.

ivbond replied:
Roles and authorities are similar in Spring. The main difference is that, roles have special semantics – starting with Spring Security 4, the ‘ROLE_‘ prefix is automatically added (if it's not already there) by any role related method.

So hasAuthority(‘ROLE_ADMIN') is similar to hasRole(‘ADMIN') because the ‘ROLE_‘ prefix gets added automatically.

vorburger commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/openjpa/OpenJpaDialect.java:
```diff
@@ -0,0 +1,153 @@
+import org.springframework.transaction.TransactionDefinition;
+import org.springframework.transaction.TransactionException;
+
+public class OpenJpaDialect extends DefaultJpaDialect {
```
@ivan333m I understand that you had to copy/paste this from Spring sources, and I personally think that is acceptable and the right way to go in this specific case. The only thing I would do here, as above, is put a JavaDoc on this class explaining it, and put a precise link to where you took this from.

vorburger commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/openjpa/OpenJpaVendorAdapter.java:
```diff
@@ -0,0 +1,111 @@
+import org.springframework.orm.jpa.vendor.Database;
+
+
+public class OpenJpaVendorAdapter extends AbstractJpaVendorAdapter {
```
again, as above JavaDoc clarifying where this is from would be useful here IMHO

ivbond replied:
Done

vorburger commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/security/filter/TenantAwareBasicAuthenticationFilter.java:
```diff
@@ -150,7 +147,8 @@ public void doFilter(final ServletRequest req, final ServletResponse res, final
 
-            super.doFilter(req, res, chain);
+//            filterChain.doFilter(request, response);
+            super.doFilterInternal(request, response, filterChain);
```
as this is security sensitive, I would love someone to have a closer look at older vs current Spring to confirm that this `doFilterInternal` is the equivalent of `doFilter` ?

ivbond replied:
`OncePerRequestFilter.doFilter` has became final in new version of spring web so I have no choice.

vorburger commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/security/filter/TenantAwareBasicAuthenticationFilter.java:
```diff
@@ -150,7 +147,8 @@ public void doFilter(final ServletRequest req, final ServletResponse res, final
             }
 
-            super.doFilter(req, res, chain);
+//            filterChain.doFilter(request, response);
```
just remove instead of commenting out this line

ivbond replied:
Done

vorburger commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/sms/scheduler/SmsMessageScheduledJobServiceImpl.java:
```diff
@@ -285,9 +285,9 @@ public void getDeliveryReports() {
-                            boolean statusChanged = false;
+                            SmsMessage smsMessage = this.smsMessageRepository.findById(smsMessageDeliveryReportData.getId()).orElse(null);
+                            Integer statusType;
+                            boolean statusChanged;
```
Why remove this? I'm proposing to put it back, shout if I misunderstood..

```suggestion
                            boolean statusChanged = false;
```

ivbond replied:
I have moved declaration to assignment in the latest commit.

vorburger commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/sms/scheduler/SmsMessageScheduledJobServiceImpl.java:
```diff
@@ -285,9 +285,9 @@ public void getDeliveryReports() {
-                            Integer statusType = smsMessage.getStatusType();
-                            boolean statusChanged = false;
+                            SmsMessage smsMessage = this.smsMessageRepository.findById(smsMessageDeliveryReportData.getId()).orElse(null);
+                            Integer statusType;
```
Why remove this? I'm proposing to put it back, shout if I misunderstood..
```suggestion
                            Integer statusType = smsMessage.getStatusType();;
```

ivbond replied:
It is redundant due to `default: statusType = smsMessage.getStatusType();`. But I rallback my changes, probably it will be more readable.

vorburger commented on fineract-provider/src/main/resources/META-INF/spring/appContext.xml:
```diff
@@ -24,17 +24,17 @@
 		pre-post-annotations="enabled" />
 
-	<tx:annotation-driven />
+<!--	<tx:annotation-driven />-->
```
@ivan333m are you replacing this XML with the `@EnableTransactionManagement` annotation in `AbstractApplicationConfiguration`, is that the same thing? Great! Then just just remove this commented out XML, it's less confusing for future readers than seeing this relict from the past.

ivbond replied:
Done

vorburger replied:
Thanks, I've seen that you've now removed it completely, perfect.

vorburger commented on fineract-provider/src/main/resources/META-INF/spring/securityContext.xml:
```diff
@@ -59,31 +60,23 @@
-			class="org.springframework.security.authentication.dao.ReflectionSaltSource">
-			<beans:property name="userPropertyToUse" value="id" />
+
+		<beans:bean id="passwordEncoder" class="org.apache.fineract.infrastructure.core.domain.DefaultPasswordEncoderFactories"
```
If we have any other reviewers more familiar with Spring Security than I am currently, it could be useful to have more eyes reviewing the changes in this file (here and below) ...

vorburger commented on fineract-provider/src/main/resources/application.properties:
```diff
@@ -17,4 +17,8 @@
+spring.profiles.default=basicauth
+management.health.jms.enabled=false
+#security.enable-csrf=false
+spring.jpa.show-sql=true
```
@ivan333m won't this (`spring.jpa.show-sql=true`) lead to A LOT of extra logging? "Log Spam" at run-time in production. Would you perhaps like to leave it here but comment it out, just as something that can be useful for debugging in development, but not in production?

ivbond replied:
Removed

vorburger replied:
Thanks! (I've reviewed that you have, so Resolve this one as well.)

vorburger commented on fineract-provider/src/test/java/org/apache/fineract/infrastructure/core/boot/tests/AbstractSpringBootWithMariaDB4jIntegrationTest.java:
```diff
@@ -20,21 +20,19 @@
+@SpringBootTest(classes = ServerWithMariaDB4jApplication.Configuration.class)
 @WebAppConfiguration
 @ActiveProfiles("basicauth")
-@IntegrationTest({ "server.port=0", "management.port=0", "mariaDB4j.port=0", "mariaDB4j.dataDir=null" })
```
@ivan333m why remove this?

ivbond replied:
@IntegrationTest was deprecated as of 1.4 in favor of org.springframework.boot.test.context.SpringBootTest.

vorburger replied:
Oh OK, thanks.

vorburger commented on fineract-provider/src/test/java/org/apache/fineract/notification/StorageTest.java:
```diff
@@ -19,6 +19,7 @@
 
 package org.apache.fineract.notification;
 
+import edu.emory.mathcs.backport.java.util.Collections;
```
just regular `java.util.Collections` would be better than `edu.emory`

ivbond replied:
Done

vorburger reviewed (changes requested):
@ivan333m again I want to thank you for this contribution - this is incredible work!

I've now done a fairly detailed code review, and think this looks great, from my side. I did have a few minor suggestions and questions which perhaps you would like to take into account. 

I'm also hoping that we get a few more eyes from the community to review this - let's see.

ivbond commented:
@vorburger Thank you for the review. I will fix all the notices in the nearest time.

muskankhedia committed:
Fixed Typo errors in README.md

ivanbonzl committed:
FINERACT-724 fixes after review

ivanbonzl committed:
FINERACT-724 fixes after review

ivanbonzl committed:
FINERACT-724 fixes after review (part2)

ivanbonzl committed:
FINERACT-724 fixes after review (part2)

vorburger commented:
@ivan333m thanks for your updates! I'll re-review as soon as I have more than a quick moment like now. (BTW @awasum @conradsp @awasum @Anh3h @vishwasbabu would you also like to review this, or shall I just go ahead and merge it when it Looks Good To Me? No stress, just checking if any interest from any of you. Specifically, there is some security related changes here which I would welcome more eyes on.)

On the last build, the `integrationTest` failed (only) below, will you have a look at that? If you suspect that it's some intermittent issue which you can not reproduce locally, then just rebase this PR so that a new build runs:

```
org.apache.fineract.integrationtests.RecurringDepositTest > testRecurringDepositAccountWithPrematureClosureTypeTransferToSavings_WITH_HOLD_TAX FAILED

    java.lang.AssertionError at RecurringDepositTest.java:643
```

ivanbonzl committed:
FINERACT-724 fixes after review (minor cleanup)

ivanbonzl committed:
FINERACT-724 fixes after review (minor cleanup)

lingocoder commented:
> „...*I'm also hoping that we get a few more eyes from the community to review this - let's see*...“

This looks like a job tailor-made for [*SonarQube's Cloud-based security vulnerability analysis*](http://bit.ly/SonarSaaS).

It codifies a bunch of heuristics and static analysis to give you a sort of automated *expert security analyst*. For free!

Here's an example of the sort of vulnerabilities it detects:

* [*ProcessorHelper - HostNameVerifier*](http://bit.ly/PrcHlprVer)
* [*TrustModifier - TrustHostnameVerifier*](http://bit.ly/TrstModVer)
* [*ProcessorHelper - checkServerTrusted*](http://bit.ly/PrcHlprSvr)
* [*ProcessorHelper - checkClientTrusted*](http://bit.ly/PrcHlprClnt)
* [*TrustModifier - checkServerTrusted*](http://bit.ly/TrstModSvr)
* [*TrustModifier - checkClientTrusted*](http://bit.ly/TrstModClnt)

As you probably already know: ***Whether or not the things it reports are real-world security concerns for your particular system, is your call***. SonarQube just tells you what it detects. Then you have to determine — [*out of all the different things it reports*](http://bit.ly/FnrActSecIssues) — which ones are *false positives* and which ones are actual vulnerabilities, in the particular context of your specific system.

HTH

edcable commented:
@lingocoder that is a great suggestion. @vorburger I don't have access to the Apache org on Github to enable SonarCloud - do you have sufficient access to do so?

ivanbonzl committed:
Merge branch 'develop' of https://github.com/ivan333m/fineract into develop

ivbond commented:
> @ivan333m thanks for your updates! I'll re-review as soon as I have more than a quick moment like now. (BTW @awasum @conradsp @awasum @Anh3h @vishwasbabu would you also like to review this, or shall I just go ahead and merge it when it Looks Good To Me? No stress, just checking if any interest from any of you. Specifically, there is some security related changes here which I would welcome more eyes on.)
> 
> On the last build, the `integrationTest` failed (only) below, will you have a look at that? If you suspect that it's some intermittent issue which you can not reproduce locally, then just rebase this PR so that a new build runs:
> 
> ```
> org.apache.fineract.integrationtests.RecurringDepositTest > testRecurringDepositAccountWithPrematureClosureTypeTransferToSavings_WITH_HOLD_TAX FAILED
> 
>     java.lang.AssertionError at RecurringDepositTest.java:643
> ```

I can't reproduce the fail locally. So I have made a rebase. But build failed again. In logs I have found 
`BUILD SUCCESSFUL in 18m 8s ` and a batch of deadlocks. It a little bit confusing. :)

edcable commented:
@vorburger once we merge this, will we be able to also merge https://github.com/apache/fineract/pull/649

awasum commented:
Hi All,

Looks like the Travis result shows the health endpoint is not available. I checked locally, the build is Ok but the health endpoint which is called in Travis is not available. I even added an exception in Spring Security configuration locally and it did not solve the problem. See: 
![image](https://user-images.githubusercontent.com/15528032/69317488-5660a700-0c3b-11ea-8892-516676bf8798.png)
 
When I add basic auth credentials, I get 401: Unauthorized..See: 
![image](https://user-images.githubusercontent.com/15528032/69317559-7ee8a100-0c3b-11ea-96a0-1f8cc987e105.png)



Secondly, when I connect this upgraded version of the backend to the Mifos Front end here: https://github.com/openMF/community-app, I am unable to log in to the App with credentials mifos/password.

Does anyone have an idea why this is happening? Maybe, I am missing something.

ivanbonzl committed:
FINERACT-724 disable csrf on oauth configuration

awasum commented:
> > „..._I'm also hoping that we get a few more eyes from the community to review this - let's see_...“
> 
> This looks like a job tailor-made for [_SonarQube's Cloud-based security vulnerability analysis_](http://bit.ly/SonarSaaS).
> 
> It codifies a bunch of heuristics and static analysis to give you a sort of automated _expert security analyst_. For free!
> 
> Here's an example of the sort of vulnerabilities it detects:
> 
> * [_ProcessorHelper - HostNameVerifier_](http://bit.ly/PrcHlprVer)
> * [_TrustModifier - TrustHostnameVerifier_](http://bit.ly/TrstModVer)
> * [_ProcessorHelper - checkServerTrusted_](http://bit.ly/PrcHlprSvr)
> * [_ProcessorHelper - checkClientTrusted_](http://bit.ly/PrcHlprClnt)
> * [_TrustModifier - checkServerTrusted_](http://bit.ly/TrstModSvr)
> * [_TrustModifier - checkClientTrusted_](http://bit.ly/TrstModClnt)
> 
> As you probably already know: _**Whether or not the things it reports are real-world security concerns for your particular system, is your call**_. SonarQube just tells you what it detects. Then you have to determine — [_out of all the different things it reports_](http://bit.ly/FnrActSecIssues) — which ones are _false positives_ and which ones are actual vulnerabilities, in the particular context of your specific system.
> 
> HTH

Looks like Apache Infra has a process to enable this: https://cwiki.apache.org/confluence/display/INFRA/SonarQube+Analysis
We might look at this in a separate PR/issue.

awasum commented:
@ivan333m , @vorburger . I think the Health Actuator endpoint changed in recent version of spring boot(2.1.7) which this PR is using. So in Travis, we might need to change the http test call to: https://localhost:8443/fineract-provider/actuator/health. This should hopefully solve the failing Travis test. @ivan333m will you be able to help with this? Here is the travis file/line: https://github.com/apache/fineract/blob/23c3f23e12ad04563b2c0d5353b143cdf70911de/.travis.yml#L65

I would have done a PR but I like it to be done here so the review comments are in one place...I might send a test PR to see if am right, but go ahead and make the change here on your branch.

ivanbonzl committed:
FINERACT-724 fix travis build after spring version migration

ivbond commented:
Thank you, Awasum!!!

I have just committed the change to my branch.

Ivan.

чт, 21 нояб. 2019 г. в 12:22, Awasum Yannick <notifications@github.com>:

> @ivan333m <https://github.com/ivan333m> , @vorburger
> <https://github.com/vorburger> . I think the Health Actuator endpoint
> changed in recent version of spring boot(2.1.7) which this PR is using. So
> in Travis, we might need to change the http test call to:
> https://localhost:8443/fineract-provider/actuator/health. This should
> hopefully solve the failing Travis test. @ivan333m
> <https://github.com/ivan333m> will you be able to help with this? Here is
> the travis file/line:
> https://github.com/apache/fineract/blob/23c3f23e12ad04563b2c0d5353b143cdf70911de/.travis.yml#L65
>
> I would have done a PR but I like it to be done here so the review
> comments are in one place...I might send a test PR to see if am right, but
> go ahead and make the change here on your branch.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/fineract/pull/642?email_source=notifications&email_token=AMQZIK3Z5VB26XTPZTPZRRLQUZOPZA5CNFSM4I7ZF742YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEEZW6LA#issuecomment-557018924>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMQZIK7VYPXKTRP2VOO3WYDQUZOPZANCNFSM4I7ZF74Q>
> .
>

ivanbonzl committed:
FINERACT-724 fix oauth configuration

edcable commented:
@awasum @ivan333m I opened a ticket on Apache Infra for sonarcloud.io - it's now there for Apache Fineract and needs to be configured. I'm waiting for them to enable it across the Fineract CN repos - would it be all of them?

edcable commented:
If you log into sonarcloud.io with your apache linked github ID you should be able to access the Apache Fineract instance to configure.

awasum commented:
> @awasum @ivan333m I opened a ticket on Apache Infra for sonarcloud.io - it's now there for Apache Fineract and needs to be configured. I'm waiting for them to enable it across the Fineract CN repos - would it be all of them?

Yes, I think.

ivbond commented:
I have found that the upgrade has broken `oauth` authentication.
I will try to fix this issue.

awasum commented:
> @ivan333m again I want to thank you for this contribution - this is incredible work!
> 
> I've now done a fairly detailed code review, and think this looks great, from my side. I did have a few minor suggestions and questions which perhaps you would like to take into account.
> 
> I'm also hoping that we get a few more eyes from the community to review this - let's see.

I have done a quick review of this PR. It looks good to me. I have connected this upgraded backend with mifos community app and tested that login works. Most major CRUD operation works as far as I can tell. I have not done a security audit to see how the upgrades affects security. We can do that later. @vorburger I wont know if this resolves the conversation for you as your review here is the only red flag preventing me from merging this.

awasum commented:
> I have found that the upgrade has broken `oauth` authentication.
> I will try to fix this issue.

Ok..Go ahead.

ivanbonzl committed:
FINERACT-724 fix oauth authentication, replace DefaultPasswordEncoderFactories with standard spring security PasswordEncoderFactories

ivbond commented:
I have committed fix for oauth authentication and some refactoring for
basic.

It would be nice additionally check oauth authentication (I have made some
checks, but due to lack of integration tests, it should be checked twice)

I think the PR could be merged if there are no probs with these cases.

@Michael, please, let me know if I miss (leave unanswered) the points from
your review.

Best regard,
Ivan


пт, 22 нояб. 2019 г., 13:44 Awasum Yannick <notifications@github.com>:

> I have found that the upgrade has broken oauth authentication.
> I will try to fix this issue.
>
> Ok..Go ahead.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/fineract/pull/642?email_source=notifications&email_token=AMQZIK6M5HO2Y62XE6BYADLQU7A2XA5CNFSM4I7ZF742YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEE5MZBQ#issuecomment-557501574>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMQZIKZYZD5WXZJ3NSM5ULTQU7A2XANCNFSM4I7ZF74Q>
> .
>

vorburger commented:
Hello! I'm alive, was just really busy at work.. I'll re-review and
hopefully (likely) merge this some time over this weekend..

>

vorburger reviewed (commented):
This looks great now! The only loose end is that somehow the current 31 Commits and 206 Files changed look wrong... this seems to include a lot of existing older things from others which are already on `develop`, that seems weird.  I'll see if I can fix that with a `git rebase` on this or a new PR (preserving your authorship).

vorburger commented:
Hm, this isn't trivial to clean up... I'll have another go at this may be tomorrow or Monday night.  @awasum  or @ivan333m if you can figure out how to make this have a cleaner commit and change list, go for it...

vorburger commented:
FTR: This was merged in #653.
